GridPlus/phonon-website
# Add GRID -> PHONON redemption support (configured for Rinkeby!)

⚠️ **This code is currently configured for Rinkeby!** DO NOT deploy to production without changing essential configuration values (see code comments). ⚠️ 

----------------

This PR adds the ability for users to redeem GRID tokens for PHONON. 

#### Couple things to note: 

- This PR is currently configured for Rinkeby. The constants within App.tsx will need to be edited before deploying to production (after deploying the production contracts, of course)
- On Rinkeby, we use the DAI and USDC contracts in place of GRID and PHONON, respectively, because those contracts already exist on Rinkeby. The code follows the ERC 20 spec so they should work just fine with GRID and PHONON when deployed to production. 
- To test this, you'll need to acquire some DAI to redeem (and potentially USDC to fill up the redemption contract) on Rinkeby. Follow the instructions below.
- The happy path of this redemption flow assumes you have >1 DAI (GRID) and 0 USDC (PHONON). Note that you'll get 155x the USDC, so it's best to use small amounts in order for the redemption to succeed. Also note that the redemption contract itself needs to own the USDC to be redeemed, as it's the contract that sends it to the user.   

#### Some notes on quality: 

- I'd call the code quality here "good", although I've cut some corners for time. I've added comments wherever this occurred, and TODO's where applicable. 
- The UX should be good enough, unless there's a use case I'm not thinking about. That said, I left the UI ugly on purpose for @douglance to clean up and change to his liking. Aside: Don't forget to style the error message that sometimes can occur, for instance, when a user rejects a transaction from their wallet.
- I didn't follow the style of the initial code as it pertains to displayed strings (that is, I didn't add displayed strings into a json file like the rest of the code). @douglance I figured this was best for you to organize how you saw fit, especially if you wanted to add/remove anything from the UI/UX. 

#### Wallets supported:

- Metamask (Desktop)
- Frame (Desktop)

Currently there is no support for mobile wallets. To do so, we'll need an Infura ID so we can add Wallet Connect, as is required by their integration. 

Note that we can easily add support for any wallet supported by [web3modal](https://github.com/Web3Modal/web3modal). if we *do* add a third wallet (say, Wallet Connect), the user will receive a pop up when clicking on the "Connect Wallet" button, vs. the auto-connect behavior that happens now. 

#### Instructions to test on Rinkeby: 

To test out this feature on Rinkeby, follow the steps below:

1. Use the Compound faucet at the following link on Rinkeby to acquire some DAI: https://app.compound.finance/ (see video in discord)
2. Transfer all but 1 DAI into another wallet you control. We do this so we don't exhaust the test Redeemer's store of USDC, given the 155x multiplier.
3. Before continuing, make sure that the Redeemer contract has at least 155 USDC. If not, use the same faucet to get USDC, and transfer that USDC to the Redeemer contract. Note, you may need to do this twice, as the faucet send USDC in 100 USDC allotments.
4. Open the page, connect your wallet using the Connect Wallet button. Make sure you're using the wallet address that has exactly 1 DAI. You should see your wallet address shown, and your DAI balance displayed as "GRID". 
5. A button should appear telling you that you must first approve PHONON to convert your GRID. Click this button to fire off the ERC-20 approval transaction.
6. When that transaction succeeds, a new button should appear telling you that you can redeem your GRID for a new amount of PHONON. Click this button to fire off a transaction.
6. Once that transaction is successful, you'll see a message saying that the redemption succeeded. If it wasn't successful, or anything has gone wrong throughout the process, an error will be shown.
